lib/typed/coercion/struct_coercer.rb

```diff
@@ -21,7 +21,41 @@ def coerce(type:, value:)
 
         return Failure.new(CoercionError.new("Value of type '#{value.class}' cannot be coerced to #{type} Struct.")) unless value.is_a?(Hash)
 
-        Success.new(T.cast(type, T::Types::Simple).raw_type.from_hash!(HashTransformer.new.deep_stringify_keys(value)))
+        values = {}
+
+        type = T.cast(type, T::Types::Simple)
+
+        type.raw_type.props.each do |name, prop|
+          attribute_type = prop[:type_object]
+          value = HashTransformer.new.deep_symbolize_keys(value)
+
+          if value[name].nil?
+            # if the value is nil but the type is nilable, no need to coerce
+            next if attribute_type.respond_to?(:valid?) && attribute_type.valid?(value[name])
+
+            return Typed::Failure.new(CoercionError.new("#{name} is required but nil given"))
+          end
+
+          # now that we've done the nil check, we can unwrap the nilable type to get the raw type
+          simple_attribute_type = attribute_type.respond_to?(:unwrap_nilable) ? attribute_type.unwrap_nilable : attribute_type
+
+          # if the prop is a struct, we need to recursively coerce it
+          if simple_attribute_type.respond_to?(:raw_type) && simple_attribute_type.raw_type.respond_to?(:props)
+            Typed::HashSerializer
+              .new(schema: Typed::Schema.from_struct(simple_attribute_type.raw_type))
```

Comment:
suggestion(non-blocking): If we know this is a struct, then we should be able to call the `schema` method we're adding to them. Further, instead of checking if it responds to props, maybe we do a cast check to help Sorbet out?

```suggestion
              .new(schema: simple_attribute_type.raw_type.schema)
```

Reply:
👍 updated
[b714824](https://github.com/maxveldink/sorbet-schema/pull/102/commits/b71482490cb1f7c579891c38b2212b8568b7909e) to use `.schema`
[7662eff](https://github.com/maxveldink/sorbet-schema/pull/102/commits/7662effef962c3c1529016dea353cd62bc065ce2) to check type...I _think_ this is what you meant but let me know if you were thinking of a different way